legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/concurrent/ConcurrentExecutionNodeExecutorPool.java

```diff
@@ -0,0 +1,158 @@
+import io.opentracing.util.GlobalTracer;
+import org.eclipse.collections.api.list.MutableList;
+import org.eclipse.collections.api.tuple.Pair;
+import org.eclipse.collections.impl.factory.Lists;
+import org.eclipse.collections.impl.list.mutable.FastList;
+import org.eclipse.collections.impl.tuple.Tuples;
+import org.eclipse.collections.impl.utility.ListIterate;
+import org.finos.legend.engine.plan.execution.nodes.ExecutionNodeExecutor;
+import org.finos.legend.engine.plan.execution.nodes.state.ExecutionState;
+import org.finos.legend.engine.plan.execution.result.ConstantResult;
+import org.finos.legend.engine.plan.execution.result.Result;
+import org.finos.legend.engine.protocol.pure.v1.model.executionPlan.nodes.ExecutionNode;
+import org.finos.legend.engine.shared.core.operational.opentracing.TracedSupplier;
+import org.finos.legend.engine.shared.core.url.StreamProvider;
+import org.finos.legend.engine.shared.core.url.StreamProviderHolder;
+import org.pac4j.core.profile.CommonProfile;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class ConcurrentExecutionNodeExecutorPool
+{
+    private static final AtomicReference<ConcurrentExecutionNodeExecutorPool> INSTANCE = new AtomicReference<>();
+
+    private final int poolSize;
+    private final ThreadPoolExecutor executor;
+    private final Semaphore availableThreads;
+
+    private ConcurrentExecutionNodeExecutorPool(int poolSize)
+    {
+        this.poolSize = poolSize;
+        this.executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(poolSize);
+        this.availableThreads = new Semaphore(poolSize);
+    }
+
+    public static void initializeExecutorPool(int poolSize)
+    {
+        boolean success = INSTANCE.compareAndSet(null, new ConcurrentExecutionNodeExecutorPool(poolSize));
+        if (!success)
+        {
+            throw new UnsupportedOperationException("Can't initialize ConcurrentExecutionNodeExecutorPool multiple times");
+        }
+    }
+
+    public static void teardownExecutorPool()
+    {
+        ConcurrentExecutionNodeExecutorPool concurrentExecutionNodeExecutorPool = getExecutorPool();
+
+        if (concurrentExecutionNodeExecutorPool == null)
+        {
+            throw new IllegalStateException("teardown called on un-initialized pool");
+        }
+
+        concurrentExecutionNodeExecutorPool.executor.shutdown();
+
+        INSTANCE.set(null);
+    }
+
+    public static ConcurrentExecutionNodeExecutorPool getExecutorPool()
+    {
+        return INSTANCE.get();
+    }
+
+    public List<? extends Result> execute(final List<ExecutionNode> nodes, final MutableList<CommonProfile> profiles, final ExecutionState executionState)
+    {
+        if (INSTANCE.get() != null && availableThreads.tryAcquire(nodes.size()))
```

Comment:
why do an INSTANCE.get() here, when you already inside the pool instance?

Reply:
Moved away from Singleton pattern so no longer need this